src/main/cpp/include/genomicsdb/genomicsdb_iterators.h

```diff
@@ -0,0 +1,310 @@
+    const SingleCellTileDBIterator& operator++();
+    //Get field pointer, length, size
+    inline const uint8_t* get_field_ptr_for_query_idx(const int query_idx) const
+    {
+      assert(static_cast<size_t>(query_idx) < m_fields.size());
+      auto& genomicsdb_columnar_field = m_fields[query_idx];
+      size_t index = 0ul;
+      auto buffer_ptr = get_buffer_pointer_and_index(query_idx, index);
+      return genomicsdb_columnar_field.get_pointer_to_data_in_buffer_at_index(
+          buffer_ptr, index);
+    }
+    inline int get_field_length(const int query_idx) const
+    {
+      assert(static_cast<size_t>(query_idx) < m_fields.size());
+      auto& genomicsdb_columnar_field = m_fields[query_idx];
+      size_t index = 0ul;
+      auto buffer_ptr = get_buffer_pointer_and_index(query_idx, index);
+      return genomicsdb_columnar_field.get_length_of_data_in_buffer_at_index(
+          buffer_ptr, index);
+    }
+    inline size_t get_field_size_in_bytes(const int query_idx) const
+    {
+      assert(static_cast<size_t>(query_idx) < m_fields.size());
+      auto& genomicsdb_columnar_field = m_fields[query_idx];
+      size_t index = 0ul;
+      auto buffer_ptr = get_buffer_pointer_and_index(query_idx, index);
+      return genomicsdb_columnar_field.get_size_of_data_in_buffer_at_index(
+          buffer_ptr, index);
+    }
+    inline bool is_valid(const int query_idx) const
+    {
+      assert(static_cast<size_t>(query_idx) < m_fields.size());
+      auto& genomicsdb_columnar_field = m_fields[query_idx];
+      size_t index = 0ul;
+      auto buffer_ptr = get_buffer_pointer_and_index(query_idx, index);
+      return genomicsdb_columnar_field.is_valid_data_in_buffer_at_index(
+          buffer_ptr, index);
+    }
+    inline const int64_t* get_coordinates() const
+    {
+      auto coords_query_idx = m_fields.size()-1u;
+      return reinterpret_cast<const int64_t*>(get_field_ptr_for_query_idx(coords_query_idx));
+    }
+    void print(const int query_idx, std::ostream& fptr=std::cout) const;
+    void print_ALT(const int query_idx, std::ostream& fptr=std::cout) const;
+    void print_csv(const int query_idx, std::ostream& fptr=std::cout) const;
+    inline bool end() const { return m_done_reading_from_TileDB && m_PQ_live_cell_markers.empty(); }
+  protected:
+    /*
+     * Does one read for the attributes in m_query_attribute_idx_vec
+     * The first call must pass the 3 arguments correctly
+     * Subsequent calls can skip the arguments
+     */
+    void read_from_TileDB(TileDB_CTX* tiledb_ctx=0, const char* array_path=0,
+        std::vector<const char*>* attribute_names=0);
+    //Given the columnar field object, get the buffer pointer and index in the buffer
+    //Response depends on whether this is a cell that begins before the query interval begin
+    //or cell >= query interval begin
+    inline const GenomicsDBBuffer* get_buffer_pointer_and_index(const int field_query_idx, size_t& index) const
```

Comment:
times like this I wish C++ had a Python way of easily returning multiple values from a method call :). Maybe you can consider returning a std::pair. This works though